src/libcollections/btree/map.rs

```diff
@@ -687,63 +687,59 @@ impl<K: Ord, V> BTreeMap<K, V> {
-                            Internal(internal) => first_leaf_edge(internal.descend()),
-                        }
-                    }
-                    GoDown(bottom) => bottom,
-                }
-            }
-            Unbounded => first_leaf_edge(self.root.as_ref()),
-        };
+        let mut min_node = self.root.as_ref();
+        let mut max_node = self.root.as_ref();
+        let mut min_found = false;
+        let mut max_found = false;
+        let mut diverged = false;
 
-        let back = match max {
-            Included(key) => {
-                match search::search_tree(self.root.as_ref(), key) {
-                    Found(kv_handle) => {
-                        match kv_handle.right_edge().force() {
-                            Leaf(bottom) => bottom,
-                            Internal(internal) => first_leaf_edge(internal.descend()),
-                        }
-                    }
-                    GoDown(bottom) => bottom,
-                }
-            }
-            Excluded(key) => {
-                match search::search_tree(self.root.as_ref(), key) {
-                    Found(kv_handle) => {
-                        match kv_handle.left_edge().force() {
-                            Leaf(bottom) => bottom,
-                            Internal(internal) => last_leaf_edge(internal.descend()),
-                        }
-                    }
-                    GoDown(bottom) => bottom,
-                }
+        loop {
+            let min_edge = match (min_found, range.start()) {
+                (false, Included(key)) => match search::search_linear(&min_node, key) {
+                    (i, true) => { min_found = true; i },
+                    (i, false) => i,
+                },
+                (false, Excluded(key)) => match search::search_linear(&min_node, key) {
+                    (i, true) => { min_found = true; i+1 },
+                    (i, false) => i,
+                },
+                (_, Unbounded) => 0,
+                (true, Included(_)) => min_node.keys().len(),
+                (true, Excluded(_)) => 0,
+            };
+
+            let mut max_edge = match (max_found, range.end()) {
+                (false, Included(key)) => match search::search_linear(&max_node, key) {
+                    (i, true) => { max_found = true; i+1 },
+                    (i, false) => i,
+                },
+                (false, Excluded(key)) => match search::search_linear(&max_node, key) {
+                    (i, true) => { max_found = true; i },
+                    (i, false) => i,
+                },
+                (_, Unbounded) => max_node.keys().len(),
+                (true, Included(_)) => 0,
+                (true, Excluded(_)) => max_node.keys().len(),
+            };
+
+            if !diverged {
+                // Don't allow max_edge to go to the left of min_edge
+                if max_edge < min_edge { max_edge = min_edge; }
```

Comment:
If we do go this route of never panicking, this also could be an optimization - just search for `max_edge` starting from `min_edge`. This would never produced "crossed nodes" by construction, and would require fewer comparisons. I don't know whether this is worth it though. Additionally, it can be done later.